### PR TITLE
python: Don't prefer 3.10

### PR DESCRIPTION
**Pull request by jmcarcell**
3.11 has been out for some time; the preference was introduced last November when 3.11 was added for the first time.

### CODE REVIEW

**Review by adamjstewart — DISMISSED** [review]
Looks like both conda and homebrew now default to 3.11 as well, should be stable at this point.

**Comment by haampie** [comment]
I agree, and repeat once more that we should stop preferring old Python, and instead set upperbounds on packages that don't support it :upside_down_face:.

I'll push a few upperbounds to this PR to make CI pass.

**Comment by haampie** [comment]
@jthies I've created an issue upstream about lacking 3.11 support of phist, the comments about 3.11 support in the spack package seem wrong?

**Review by haampie — APPROVED** [review]

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
> I agree, and repeat once more that we should stop preferring old Python, and instead set upperbounds on packages that don't support it 🙃.

I'm wary of this, especially because of what's coming in 3.12 (dropping distutils). We will find that most older packages (scientific software that hasn't had a new release in over a decade but is still critical) won't build anymore and someone (me) will have to deal with all the tickets. I would be much happier about preferring 3.12 if every Python package was built in CI.

File: var/spack/repos/builtin/packages/phist/package.py [hunk 1]
```diff
@@ -175,7 +173,8 @@ class Phist(CMakePackage):
     # Python 3 or later is required for generating the Fortran 2003 bindings
     # since version 1.7, you can get rid of the dependency by switching off
     # the feature (e.g. use the '~fortran' variant)
-    depends_on("python@3:", when="@1.7: +fortran", type="build")
+    # For the upperbound see https://bitbucket.org/essex/phist/issues/246/does-not-build-with-python-311
+    depends_on("python@3:3.10", when="@1.7: +fortran", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/phist/package.py:177]: Do we have a preference for depends_on vs. conflicts? It used to be the case that depends_on guided the concretizer and conflicts crashed it, but that's no longer the case for clingo. I usually use depends_on for what is documented as a dep and conflicts for surprises we find (aka bugs).

File: var/spack/repos/builtin/packages/py-protobuf/package.py [hunk 2]
```diff
@@ -62,6 +62,8 @@ class PyProtobuf(PythonPackage):
     version("3.3.0", sha256="1cbcee2c45773f57cb6de7ee0eceb97f92b9b69c0178305509b162c0160c1f04")
     version("3.0.0", sha256="ecc40bc30f1183b418fe0ec0c90bc3b53fa1707c4205ee278c6b90479e5b6ff5")
 
+    # newer versions support python 3.11+, but we need to switch to bazel
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-protobuf/package.py:65]: This may or may not be true: #38614

**Pull request merged**